### ISSUE TITLE
[BUG] wash-lib should be buildable with any combination of feature flags

### ISSUE THREAD

**Issue by connorsmith256** [wasmCloud/wasmCloud#862]
**Describe the bug**
wash-lib cannot be built with `--no-default-features`

**To Reproduce**
1. `cd crates/wash-lib`
2. `cargo build --no-default-features`

Note that the library doesn't compile

**Comment by rimbi** [comment]
Just to double check: Is the assumption that `cargo build --no-default-features` should succeed correct?

**Comment by brooksmtownsend** [comment]
> Is the assumption that cargo build --no-default-features should succeed correct?

Yes that's right, the library should be buildable without any features

**Comment by rimbi** [comment]
There seems to be 2 feature flags which cause this problem: `async-nats` and `parser`. Should  any of them by any chance be part of the `no-default-features` set? If not, can you please explain why? @brooksmtownsend

**Comment by thomastaylor312** [comment]
I would need to look through again but I feel like serde is used everywhere and shouldn't be behind a flag. I also think async nats is used everywhere, but there could be some of the things like signing or pulling that wouldn't necessarily need that feature enabled. I think I'd rather just have those not behind a feature flag, and if anyone ever needs us to slim it down, we can put it behind a different feature flag. Thoughts @brooksmtownsend?

**Comment by brooksmtownsend** [comment]
Yeah I have a hard time trying to consider separating out `serde` and `serde_json`, nearly all of our structures have the Serialize and Deserialize derive traits. `async_nats` I can see being more useful to have behind that `nats` flag, but without a clear goal (so we can compile wash-lib to a Wasm component, for example) it's not a requirement in my eyes. 

To be clear, serde/serde_json should be a part of `wash-lib` by default, no flag to disable it, unless we really find a clear use case for that in the future. I like gating the NATS libraries/features behind a flag so they can be disabled

**Comment by rimbi** [comment]
Thanks for the explanation @thomastaylor312 @brooksmtownsend . Apparently, async_nats Client is used as the only means for the RPC connection without an alternative. Does it make sense to disable it by gating behind a feature flag?

**Comment by brooksmtownsend** [comment]
@rimbi yes I think so, right now we gate the `app` module behind the NATS flag and we should disable the `async_nats` dependency if the `nats` feature isn't provided. We may need to gate other modules behind the `nats` flag in order for this to work but it's a good idea.

**Comment by rimbi** [comment]
@brooksmtownsend the following changes remove the build problem but I am not sure if it makes sense. Any opinion?
```
index a9597ce5..07f56f83 100644
--- a/crates/wash-lib/src/lib.rs
+++ b/crates/wash-lib/src/lib.rs
@@ -28,14 +28,21 @@ pub mod parser;
 #[cfg(feature = "start")]
 pub mod start;
 
+#[cfg(feature = "nats")]
 pub mod actor;
+#[cfg(feature = "nats")]
 pub mod capture;
 pub mod common;
+#[cfg(feature = "nats")]
 pub mod config;
+#[cfg(feature = "nats")]
 pub mod context;
+#[cfg(feature = "nats")]
 pub mod drain;
 pub mod id;
 pub mod keys;
 pub mod registry;
+#[cfg(feature = "nats")]
 pub mod spier;
+#[cfg(feature = "nats")]
 pub mod wait;
```

**Comment by thomastaylor312** [comment]
I think the fact that it solves the build problem is good enough for now! We can always come back and make the features better

**Comment by thomastaylor312** [comment]
Closed in #1525
